### PR TITLE
DOCKER-523: implement docker commit

### PR DESCRIPTION
**Pull request by twhiteman**

### CODE REVIEW

File: bin/docker-build.js [hunk 1]
```diff
@@ -57,9 +57,10 @@ process.on('message', function (message) {
     assert.optionalNumber(message.timeoutSeconds, 'timeoutSeconds');
 
     // Setup log streams.
+    var commandType = message.payload.command;  // Either 'build' or 'commit'
```
Comment (misterdjules) [inline bin/docker-build.js]: Should `commandType` be tested for being either `'build'` or `'commit'` and error if anything else?

Comment (misterdjules) [inline bin/docker-build.js]: Same in `setupDockerBuildSocket`.

File: bin/docker-build.js [hunk 2]
```diff
@@ -453,25 +471,183 @@ function buildFromContext(opts, callback) {
...
+
+/**
+ * Create sdc-docker-build Builder instance and manage the image create process.
+ *
+ * Communicates (relay) between sdc-docker and the image Builder.
+ */
+function commitImage(opts, callback) {
+    assert.object(opts.payload, 'opts.payload');
+    assert.object(opts.payload.fromImage, 'opts.payload.fromImage');
+    assert.string(opts.payload.fromImageUuid, 'opts.payload.fromImageUuid');
```
Comment (misterdjules) [inline bin/docker-build.js]: Should `opts.payload.changes` be asserted too?

File: bin/docker-build.js [hunk 3]
```diff
@@ -453,25 +471,183 @@ function buildFromContext(opts, callback) {
...
+
+    var changes = opts.payload.changes;
+    var fromImage = opts.payload.fromImage;
+    var fromImageUuid = opts.payload.fromImageUuid;
+    var log = opts.log;
+    var pendingCallbackEvents = {};
+    var socket = opts.buildSocket;
+
+    socket.on('end', function () {
```
Comment (misterdjules) [inline bin/docker-build.js]: We may want to add listeners for the `'error'` and `'timeout'` events.

File: bin/docker-build.js [hunk 4]
```diff
@@ -453,25 +471,183 @@ function buildFromContext(opts, callback) {
...
+                }
+                break;
+        }
+
+        if (event.callback) {
+            pendingCallbackEvents[event.messageId] = event;
+        }
+    });
+
+    builder.on('end', function (err) {
```
Comment (misterdjules) [inline bin/docker-build.js]: Does `dockerbuild.Builder` instances emit `'error'` events, and if so should we add a listener for these?

File: bin/docker-build.js [hunk 5]
```diff
@@ -860,9 +1043,16 @@ function importImageSnapshotsIntoImgapi(builder, opts, callback) {
...
             if (err) {
-                log.error('Error importing images into IMGAPI', err);
+                log.error('Unable to import image into imgapi: %s', err);
                 // Remove this image if it failed to import/validate.
+                if (!imageSdcDocker) {
+                    cb(err);
+                    return;
+                }
                 var image_uuid = imageSdcDocker.image_uuid;
+                log.error('Unable to import image %s into imgapi: %s',
```
Comment (misterdjules) [inline bin/docker-build.js]: Isn't that message a duplicate of [the one above](https://github.com/joyent/sdc-cn-agent/pull/12/files#diff-5a31125d81398aadaf73b9ba01aeb9a5R1046)?

File: bin/docker-build.js [hunk 6]
```diff
@@ -757,8 +938,10 @@ function importImageSnapshotsIntoImgapi(builder, opts, callback) {
         }
 
         log.debug('Creating image for build layer %d %j', idx, dockerImage);
-        builder.emitStdout(util.format('Importing image %s into '
-            + 'IMGAPI\n', builder.getShortId(dockerImage.id)));
+        if (builder.commandType !== 'commit') {
```
Comment (misterdjules) [inline bin/docker-build.js]: Why don't we output this message in the case of commit?

File: bin/docker-build.js [hunk 7]
```diff
@@ -938,15 +1129,20 @@ function zfsSnapshotStream(opts, callback, onProcessError) {
...
     var args = [
         '-r', 'root',                         // Include zone root directory
         '-x', 'native',                       // Exclude anything in /native/
         '-x', 'checkpoints',                  // Exclude snapshot /checkpoints/
         '-x', 'var/svc/provision_success',    // Exclude vm provisioning file
-        '-x', 'var/log/sdc-dockerinit.log',   // Exclude sdc docker log file
-        zfsBase, parent_snapshot, snapshot
+        '-x', 'var/log/sdc-dockerinit.log'    // Exclude sdc docker log file
     ];
+    if (opts.commandType === 'commit') {
```
Comment (misterdjules) [inline bin/docker-build.js]: I'm ignorant of the reason why we differentiate these two cases, maybe a comment would help any other reader not familiar with this code base :)

File: bin/docker-build.js (hunk 1 above)
Comment (twhiteman) [inline bin/docker-build.js]: Yes - it should.

File: bin/docker-build.js (hunk 2 above)
Comment (twhiteman) [inline bin/docker-build.js]: Indeed.

File: bin/docker-build.js (hunk 4 above)
Comment (twhiteman) [inline bin/docker-build.js]: No 'error' events, errors should all be passed back through the callbacks.

File: bin/docker-build.js (hunk 5 above)
Comment (twhiteman) [inline bin/docker-build.js]: No way - that one is totally different ;)

I had meant to use that for the case where there was imageSdcDocker (and thus no image_uuid) - I'll move that log into the `if (!imageSdcDocker) {` block.

File: bin/docker-build.js (hunk 6 above)
Comment (twhiteman) [inline bin/docker-build.js]: No stdout messages should be generated for a docker commit... and this check should be moved to the actual `build.emitStdout` function.

File: bin/docker-build.js (hunk 7 above)
Comment (twhiteman) [inline bin/docker-build.js]: Will add.

**Pull request closed without merging**